### DISCUSSION TITLE
Custom serializer for plist inside XML

### DISCUSSION THREAD

**Discussion by adrientetar** [pdvrieze/xmlutil#153] [Q&A]
I want to parse an XML file with plist embedded in it, like the `<dict>` element below:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<glyph name="space" format="2">
  <lib>
    <dict>
      <key>public.markColor</key>
      <string>0.97,1,0,1</string>
    </dict>
  </lib>
</glyph>
```

I would like to deserialize the dict using a separate library (dd-plist). I wrote a custom serializer to NSDictionary, however I can't figure out how to tell the Decoder to decode an xml tag to a string. (It seems like I can't cast the Decoder to XmlDecoder as that's an internal type?)

```kotlin
class NSDictionarySerializer : KSerializer<NSDictionary> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("dict", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): NSDictionary =
        PropertyListParser.parse(
            // wrap the element in a proper doctype for parsing
            buildString {
                append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
                    "<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n" +
                    "<plist version=\"1.0\">\n")
                append(decoder.decodeString())
                append("\n</plist>")
            }.byteInputStream()
        ).toDictionary()

    override fun serialize(encoder: Encoder, value: NSDictionary) {
        val outputStream = ByteArrayOutputStream()
        XMLPropertyListWriter.write(value, outputStream)
        val plist = outputStream.toString(Charset.forName("UTF-8"))
        val dictString = buildString {
            // remove doctype etc. to just have the element
            val content = plist.substringAfter("<dict>")
                .substringBeforeLast("</dict>")

            append("<dict>")
            if (content.isNotBlank()) {
                append(content)
            }
            append("</dict>")
        }
        encoder.encodeString(dictString)
    }
}
```

I would appreciate any pointer; thanks for this library!

**Comment by pdvrieze** [comment]
First of all, the types to use for specialisation in custom serializers are the interfaces `XmlInput` and `XmlOutput` respectively. Due to the way that xml parsers work it is not possible to directly get the content of a tag as text. However in this particular case the easiest way may be to realise that the `CompactFragment` is more or less what you need (unless you want to parse plist directly using serialization - which looking at plist still requires custom serializers).


`Compactfragment`s (or the `XmlReader.siblingsToFragment` extension function) exactly capture the xml contents of a file into a fragment that holds both namespace declarations and Xml content. You can access the raw xml content as text (you may want to use `CombiningReader` to send everything into the custom `PropertyListParser`.

**Reply by pdvrieze** [reply]
An alternative that might work (depending on your parsing library) is to use the support for parsing/serializing DOM instead. It is also possible to get the underlying xml parser/serializer and use it directly (but be aware that this may break things)
